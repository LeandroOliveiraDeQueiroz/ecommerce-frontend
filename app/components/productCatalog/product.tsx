import { HeartICon } from "~/icons/heartIcon/heartICon";
import type { IFavoriteProduct } from "~/types";

export function Product({ title, price, id, image, isFavorite, onFavorite }: IFavoriteProduct) {
    const fill = isFavorite ? "var(--color-red-500)" : "none";
    const handleHeartClick = () => {
        onFavorite(id, !!isFavorite);
    }

    return (
        <div className="relative shadow-sm p-1 bg-white border border-gray-200 rounded-lg">
            <div className="absolute right-0 z-1">
                < HeartICon fill={fill} onClick={handleHeartClick} />
            </div>
            <div className="h-80 flex justify-center mt-2 mb-1">
                <img className="max-w-full max-h-full m-auto" src={image} />
            </div>
            <div className="space-y-0.5">
                <p>{title}</p>
                <p className="text-right font-bold mr-1.5 text-lg">${price}</p>
            </div>

        </div>
    );
}