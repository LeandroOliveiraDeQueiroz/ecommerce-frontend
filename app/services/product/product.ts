import axios, { type AxiosResponse } from "axios";
import type { IProduct } from "~/types";

const ProductService = () => {
  const getProducts = () => {
    return axios.get(`${import.meta.env.VITE_FAKE_API}?limit=20`);
  };

  const getProductsByIds = async (ids: number[]): Promise<IProduct[]> => {
    const promises: Promise<AxiosResponse<any, any>>[] = [];
    const products: IProduct[] = [];

    ids.forEach((id) => {
      const promise = axios.get(`${import.meta.env.VITE_FAKE_API}/${id}`);
      promises.push(promise);
    });

    const responses = await Promise.all(promises);

    responses.forEach((response) => {
      products.push(response.data);
    });

    return products;
  };

  return {
    getProducts,
    getProductsByIds,
  };
};

const productService = ProductService();

export default productService;
