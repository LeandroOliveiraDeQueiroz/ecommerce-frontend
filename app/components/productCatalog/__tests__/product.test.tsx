import { render } from "@testing-library/react"
import type { IFavoriteProduct } from "~/types"
import { Product } from "../product"

const product: IFavoriteProduct = {
    title: "T-shirt",
    description: "Blue oversized T-shirt of cotton",
    category: "clothes",
    price: 12,
    id: 1,
    image: "test",
    isFavorite: true,
    onFavorite: () => { }
}


it("should render as favorite product and show snapshot", () => {
    const { container } = render(<Product {...product} />);

    expect(container).toMatchSnapshot();

})


it("should render as NO favorite product and show snapshot", () => {
    const noFavoriteProduct = { ...product, isFavorite: false };
    const { container } = render(<Product {...noFavoriteProduct} />);

    expect(container).toMatchSnapshot();
})