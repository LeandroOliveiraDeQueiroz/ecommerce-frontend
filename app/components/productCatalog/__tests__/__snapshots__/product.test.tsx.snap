// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as NO favorite product and show snapshot 1`] = `
<div>
  <div
    class="relative shadow-sm p-1 bg-white border border-gray-200 rounded-lg"
  >
    <div
      class="absolute right-0 z-1"
    >
      <svg
        class="h-8 w-8 text-red-500 hover:fill-red-500 hover:cursor-pointer"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <div
      class="h-80 flex justify-center mt-2 mb-1"
    >
      <img
        class="max-w-full max-h-full m-auto"
        src="test"
      />
    </div>
    <div
      class="space-y-0.5"
    >
      <p>
        T-shirt
      </p>
      <p
        class="text-right font-bold mr-1.5 text-lg"
      >
        $
        12
      </p>
    </div>
  </div>
</div>
`;

exports[`should render as favorite product and show snapshot 1`] = `
<div>
  <div
    class="relative shadow-sm p-1 bg-white border border-gray-200 rounded-lg"
  >
    <div
      class="absolute right-0 z-1"
    >
      <svg
        class="h-8 w-8 text-red-500 hover:fill-red-500 hover:cursor-pointer"
        fill="var(--color-red-500)"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <div
      class="h-80 flex justify-center mt-2 mb-1"
    >
      <img
        class="max-w-full max-h-full m-auto"
        src="test"
      />
    </div>
    <div
      class="space-y-0.5"
    >
      <p>
        T-shirt
      </p>
      <p
        class="text-right font-bold mr-1.5 text-lg"
      >
        $
        12
      </p>
    </div>
  </div>
</div>
`;
